### PR TITLE
Implement classes to parse Yarn files

### PR DESCRIPTION
**Pull request by ejegrova**
Implement the classes stablished by the high-level code design (#309) to parse Yarn files (i.e. package.json and yarnrc.yml). 

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- n/a Docs updated (if applicable)
- n/a Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: cachi2/core/package_managers/yarn/project.py [hunk 1]
```diff
@@ -120,9 +148,12 @@ def yarn_cache(self) -> RootedPath:
...
         """
-        return NotImplemented
+        return self.source_dir.join_within_root(self.yarn_rc.cache_path)
 
     @classmethod
     def from_source_dir(cls, source_dir: RootedPath) -> "Project":
         """Create a Project from a sources directory path."""
-        return NotImplemented
+        yarn_rc = YarnRc.from_file(source_dir.join_within_root(".yarnrc.yml"))
+        package_json = PackageJson.from_file(source_dir.join_within_root("package.json"))
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/project.py:157]: The [docs](https://v3.yarnpkg.com/configuration/yarnrc) state that:
```
While Yarn will automatically find them in the parent directories, they should usually be kept at the root of your project (often your repository)
```

So that is one thing we need to account for when covering the path/symlink validation story.

File: cachi2/core/package_managers/yarn/project.py [hunk 2]
```diff
@@ -120,9 +148,12 @@ def yarn_cache(self) -> RootedPath:
...
         https://v3.yarnpkg.com/configuration/yarnrc#cacheFolder.
         """
-        return NotImplemented
+        return self.source_dir.join_within_root(self.yarn_rc.cache_path)
 
     @classmethod
     def from_source_dir(cls, source_dir: RootedPath) -> "Project":
         """Create a Project from a sources directory path."""
-        return NotImplemented
+        yarn_rc = YarnRc.from_file(source_dir.join_within_root(".yarnrc.yml"))
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/project.py:156]: The yarnrc is not mandatory, so we need to account for cases in which it's absent. Calling `yarn init` in a folder does not autogenerate it.

Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:156]: What if package.json is missing? Are we going to raise a `PackageRejected`?

File: cachi2/core/package_managers/yarn/project.py [hunk 3]
```diff
@@ -31,43 +39,56 @@ def cache_path(self) -> str:
...
+        return self._data.get("npmRegistryServer", "https://registry.yarnpkg.com")
 
     @property
     def yarn_path(self) -> Optional[str]:
         """Path to the yarn script present in this directory."""
-        return NotImplemented
+        try:
+            return self._data.get("yarnPath")
+        except KeyError:
+            log.warning(f"No yarnPath in {self._path}")
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/yarn/project.py:58]: Since yarnPath is not a mandatory key, I think logging a warning here may be misleading.

File: cachi2/core/package_managers/yarn/project.py [hunk 4]
```diff
@@ -31,43 +39,56 @@ def cache_path(self) -> str:
...
     @property
     def yarn_version(self) -> Optional[str]:
         """Yarn version used in this project.
 
         Extracted from the contents of the yarnPath configuration, if it is present.
         """
-        return NotImplemented
+        try:
+            yarn_path = self._data.get("yarnPath")
+            return re.search(r"([0-9]+\.[0-9]+\.[0-9]+)", yarn_path).group(0)
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:69]: +1

rather than trying to match npm semver, how about we just take this part of the filename as version?
```
some/path/yarn-(.*).cjs
               ^^^^
```

File: cachi2/core/package_managers/yarn/project.py (hunk 1 above)
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:157]: `source_dir.join_within_root("package.json")` does the validation right?

File: cachi2/core/package_managers/yarn/project.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:156]: > Calling yarn init in a folder does not autogenerate it.

Calling `yarn init -2` does, but I think you're right that it's not mandatory

File: cachi2/core/package_managers/yarn/project.py (hunk 3 above)
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:58]: +1

also `self._data.get()` will never raise a KeyError, IMO you can remove the whole try/except

File: cachi2/core/package_managers/yarn/project.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejegrova) [reply cachi2/core/package_managers/yarn/project.py]: nice :+1:  changed the regex

File: cachi2/core/package_managers/yarn/project.py (hunk 3 above)
Comment (ejegrova) [reply cachi2/core/package_managers/yarn/project.py]: removed

File: cachi2/core/package_managers/yarn/project.py [hunk 5]
```diff
@@ -31,43 +40,48 @@ def cache_path(self) -> str:
...
 
     @property
     def yarn_version(self) -> Optional[str]:
         """Yarn version used in this project.
 
         Extracted from the contents of the yarnPath configuration, if it is present.
         """
-        return NotImplemented
+        yarn_path = self.yarn_path
+        return re.search(r"yarn-(.*).cjs", yarn_path).group(1)
```
Comment (taylormadore) [inline cachi2/core/package_managers/yarn/project.py:65]: Won't we hit an issue here if self.yarn_path is None?

File: cachi2/core/package_managers/yarn/project.py (hunk 2 above)
Comment (ejegrova) [reply cachi2/core/package_managers/yarn/project.py:161]: Raising `PackageRejected` in PackageJson classmethod.

File: cachi2/core/package_managers/yarn/project.py (hunk 5 above)
Comment (ejegrova) [reply cachi2/core/package_managers/yarn/project.py:65]: yup

Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:65]: And also if the pattern doesn't match and re.search returns None (mypy caught both)

Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:65]: TBH I feel like this logic doesn't really belong in YarnRC but in the code for setting yarn version

File: cachi2/core/package_managers/yarn/project.py [hunk 6]
```diff
@@ -111,7 +141,10 @@ def is_zero_installs(self) -> bool:
         This is determined by the existence of a non-empty yarn cache folder. For more details on
         zero-installs, see: https://v3.yarnpkg.com/features/zero-installs.
         """
-        return NotImplemented
+        dir = self.yarn_cache
+        return Path.is_dir(dir.path) and any(
+            ".zip" in file.suffix for file in Path.iterdir(dir.path)
+        )
```
Comment (chmeliik) [inline cachi2/core/package_managers/yarn/project.py:147]:
```suggestion
        return dir.path.is_dir() and any(
            file.suffix == ".zip" for file in dir.path.iterdir()
        )
```

File: cachi2/core/package_managers/yarn/project.py (hunk 5 above)
Comment (ejegrova) [reply cachi2/core/package_managers/yarn/project.py]: ok, I'm gonna remove `yarn_version` property

File: cachi2/core/package_managers/yarn/project.py (hunk 1 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py]: I was thinking that a parent dir package.json could be accessed by yarn itself when we invoked any of the yarn commands. But with the validation to enforce that a valid package.json file exists in every project, this will not happen.

**Comment by brunoapimentel** [comment]
I am taking over this PR since Eliška's out for some time. Added unit tests and tried to cover more corner cases. I still need to do more testing with real scenarios, but it should be ready to review as is.

File: tests/unit/package_managers/yarn/test_project.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    return PackageJson.from_file(path)
+
+
+def test_parse_package_json(rooted_tmp_path: RootedPath) -> None:
+    package_json = _prepare_package_json_file(rooted_tmp_path, VALID_PACKAGE_JSON_FILE)
+    assert package_json.package_manager == "yarn@3.6.1"
+
+
+def test_paser_empty_package_json_file(rooted_tmp_path: RootedPath) -> None:
```
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply tests/unit/package_managers/yarn/test_project.py]: Fixed.

**Review by taylormadore — APPROVED** [review]
LGTM :+1:

File: cachi2/core/package_managers/yarn/project.py [hunk 8]
```diff
@@ -62,12 +65,26 @@ def registry_server_for_scope(self, scope: str) -> str:
...
+        return registry or self.registry_server
 
     @classmethod
     def from_file(cls, file_path: RootedPath) -> "YarnRc":
         """Parse the content of a yarnrc file."""
-        return NotImplemented
+        try:
+            with file_path.path.open("r") as f:
+                yarnrc_data = yaml.safe_load(f)
+        except yaml.error.MarkedYAMLError as e:
```
Comment (chmeliik) [inline cachi2/core/package_managers/yarn/project.py:78]: Should it be specifically MarkedYAMLError? There's also a generic yaml.YAMLError

File: cachi2/core/package_managers/yarn/project.py [hunk 9]
```diff
@@ -62,12 +65,26 @@ def registry_server_for_scope(self, scope: str) -> str:
...
+            raise PackageRejected(
+                "Can't parse the .yarnrc.yml file",
+                solution=f"Check the contents of the file. Parser error: {e}",
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:81]: > Might want to tell the user where the .yarnc.yml file is (relative to the repo root)

Or just log which path we're processing when we start processing it. There will probably be more cases like this

File: cachi2/core/package_managers/yarn/project.py [hunk 10]
```diff
@@ -4,10 +4,21 @@
...
 
+import yaml
+
+from cachi2.core.errors import PackageRejected
 from cachi2.core.rooted_path import RootedPath
 
+log = logging.getLogger(__name__)
+
+
+DEFAULT_CACHE_FOLDER = "./.yarn/cache"
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:19]: Good point, for the outside-RHTAP use case a user could indeed be setting `YARN_` environment variables

Even the path to .yarnrc.yml itself is configurable via env var (but this would _only_ work outside RHTAP, where the user can't set env vars) https://v3.yarnpkg.com/configuration/yarnrc#rcFilename

[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:19]: > Even the path to .yarnrc.yml itself is configurable via env var (but this would _only_ work outside RHTAP, where the user can't set env vars I think) https://v3.yarnpkg.com/configuration/yarnrc#rcFilename

Or maybe they can, I never know which part of the definition the user can edit or not https://tekton.dev/docs/pipelines/tasks/#specifying-a-step-template

[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/yarn/project.py:19]: > Good point, for the outside-RHTAP use case a user could indeed be setting `YARN_` environment variables
> 
> Even the path to .yarnrc.yml itself is configurable via env var (but this would _only_ work outside RHTAP, where the user can't set env vars I think) https://v3.yarnpkg.com/configuration/yarnrc#rcFilename

Aand as it turns out, in this particular case `YARN_CACHE_FOLDER` is a valid env variable.

File: cachi2/core/package_managers/yarn/project.py [hunk 11]
```diff
@@ -26,33 +37,25 @@ def __init__(self, path: RootedPath, data: dict[str, Any]) -> None:
         self._data = data
 
     @property
-    def cache_path(self) -> str:
+    def cache_folder(self) -> str:
```
[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/yarn/project.py:40]: > Stupid nitpick: Is "folder" an official term used by Yarn itself? Technically, folder can only store files, not subdirectories, so maybe `cache_dir` or `cachedir` might a more appropriate name. Looking at the code base, there is a precedent in gomod, so you can happily ignore this, it's just any Python primitive used for the actual underlying operation in this context would most likely use the word 'dir' in its name :).

Turns out, Yarn indeed calls it a _cacheFolder_, so disregard my comment please.

[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:40]: Yeah, the reason I changed it here was to match the actual configuration parameter.

File: cachi2/core/package_managers/yarn/project.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:19]: If we always set the `env` parameter when [calling](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/yarn/utils.py#L14) the subprocess, we will override any environment variable that is set in the user's env. So I think the solution here is being sure this parameter is at least set to an empty dict.

[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:19]: If we do want to expand our use case to also consider the environment variables that are already set, then we probably will need to have a separate parsing of `YARN_*` environment variables and always consider precedence for them instead of what's in yarnrc (see the early work in what we will need to support [here](https://github.com/brunoapimentel/cachi2/blob/set-yarn-config/cachi2/core/package_managers/yarn/project.py#L48-L115)).

I think it's definitely a valid use case, but we should probably handle that in a separate PR. In any case, the variable here is meant to represent the fallback value for when `cacheFolder` is not set in .yarnrc, so I think it is still correct.

File: cachi2/core/package_managers/yarn/project.py [hunk 12]
```diff
@@ -4,10 +4,21 @@
...
+import yaml
+
+from cachi2.core.errors import PackageRejected
 from cachi2.core.rooted_path import RootedPath
 
+log = logging.getLogger(__name__)
+
+
+DEFAULT_CACHE_FOLDER = "./.yarn/cache"
+DEFAULT_REGISTRY = "https://registry.yarnpkg.com"
```
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:20]: I think it can be set with `YARN_NPM_REGISTRY_SERVER`.

File: cachi2/core/package_managers/yarn/project.py (hunk 8 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py]: Changed it to YAMLError.

File: cachi2/core/package_managers/yarn/project.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py]: I did both. I'm not 100% happy with the error log output, but I think all the necessary information is there. 

We can probably revisit them with that logging story we have (if we ever get to it).

File: tests/unit/package_managers/yarn/test_project.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    assert project.yarn_rc._path == rooted_tmp_path.join_within_root(".yarnrc.yml")
+    assert project.package_json._path == rooted_tmp_path.join_within_root("package.json")
+
+
+def test_parse_project_folder_without_yarnrc(rooted_tmp_path: RootedPath) -> None:
+    _prepare_package_json_file(rooted_tmp_path, VALID_PACKAGE_JSON_FILE)
+
+    project = Project.from_source_dir(rooted_tmp_path)
+
+    assert not project.is_zero_installs
```
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply tests/unit/package_managers/yarn/test_project.py]: You're right. Added a new scenario for this test.

File: cachi2/core/package_managers/yarn/project.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/yarn/project.py:19]: Yeah, we either need to document it as unsupported or support it.

File: cachi2/core/package_managers/yarn/main.py [hunk 14]
```diff
@@ -32,6 +32,8 @@ def _resolve_yarn_project(project: Project, output_dir: RootedPath) -> list[Comp
     :param project: the directory to be processed.
     :param output_dir: the directory where the prefetched dependencies will be placed.
     """
+    log.info(f"Fetching the yarn dependencies at the subpath {output_dir.subpath_from_root}")
```
Comment (chmeliik) [inline cachi2/core/package_managers/yarn/main.py:35]: nitpick: unrelated to this commit

**Comment by chmeliik** [comment]
Marking docs as n/a because we have a docs story

**Review by lkolacek — APPROVED** [review]
LGTM :+1:

File: cachi2/core/package_managers/yarn/project.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py:19]: I'll file a story so we can discuss what's the best approach.

File: cachi2/core/package_managers/yarn/main.py (hunk 14 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/main.py:35]: I was wondering if we should start adding the logs on the main functions as the work goes or if we should have a separate story for it. In any case, since we already have the path for the file in that [error message](https://github.com/containerbuildsystem/cachi2/pull/338/files#diff-cbc68c73831a63c3d995893ed401dc07bf92cd8e8654b498177c465e8584ef18), I think I'll just drop this line here.

Comment (chmeliik) [reply cachi2/core/package_managers/yarn/main.py:35]: Yeah, just do it in a separate commit please since it's not related to project classes

Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/main.py:35]: I mean, not add the log line at all in this PR, and leave it for later. Unless you think it's nice to add it now, then I can add it in separate commit.

Comment (chmeliik) [reply cachi2/core/package_managers/yarn/main.py:35]: Ah, right. Yeah let's do it now, unless you can make sure that it will, at some point, happen :smile:

Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/main.py:36]: Ok, kept it in a separate commit.

**Comment by brunoapimentel** [comment]
Split the log message into a separate commit and added ids to the unit test scenarios so they can be more readable in the pytest output.

File: cachi2/core/package_managers/yarn/project.py [hunk 15]
```diff
@@ -4,10 +4,21 @@
 It also provides basic utility functions. The main logic to resolve and prefetch the dependencies
 should be implemented in other modules.
 """
+import json
+import logging
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/project.py:8]: importing logging and initializing the logger needs to be done in the next commit where you actually make use of the logging interface.

File: cachi2/core/package_managers/yarn/project.py [hunk 16]
```diff
@@ -111,7 +151,8 @@ def is_zero_installs(self) -> bool:
         This is determined by the existence of a non-empty yarn cache folder. For more details on
         zero-installs, see: https://v3.yarnpkg.com/features/zero-installs.
         """
-        return False
+        dir = self.yarn_cache
+        return dir.path.is_dir() and any(file.suffix == ".zip" for file in dir.path.iterdir())
```
Comment (eskultety) [inline cachi2/core/package_managers/yarn/project.py:155]: nitpick: I feel like there's too much going on on this line. Would you mind creating an intermediary variable for the comprehension and then using that inside this boolean expression? IMO it would make it a bit more readable. Having said that I know there's a lot of precedent in the code base so you can happily ignore my comment based on that, but I think despite Python allowing a lot in terms of single-line expressions/constructions, it IMO hinders readability in general.

**Review by eskultety — COMMENTED** [review]
A couple of nitpicks, nothing major though. Overall, the code looks fine to me, however, I can't provide comments on the general architecture/design wrt/ Yarn internals.

**Review by slimreaper35 — APPROVED** [review]

File: cachi2/core/package_managers/yarn/project.py (hunk 16 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py]: I changed the "is there any zip file" piece to a nested function, that way it's simple to figure out what's happening by reading the function name.

Other alternative is decomposing this to an if block, since the second piece is dependent on the is_dir check. Let me know if you prefer it that way.

Comment (eskultety) [reply cachi2/core/package_managers/yarn/project.py]: > Other alternative is decomposing this to an if block, since the second piece is dependent on the is_dir check. Let me know if you prefer it that way.

Call me old-fashioned, but ^this to me sounds like the more readable approach.

Comment (brunoapimentel) [reply cachi2/core/package_managers/yarn/project.py]: Ok, changed it to an if block.

**Review by eskultety — APPROVED** [review]
